bundles/config/org.eclipse.smarthome.config.discovery.usbserial/src/main/java/org/eclipse/smarthome/config/discovery/usbserial/internal/UsbSerialDiscoveryService.java

```diff
@@ -0,0 +1,179 @@
+ * are called by this discovery service whenever new devices are detected or devices are removed. Such
+ * {@link UsbSerialDiscoveryParticipant}s should be provided by bindings accessing USB devices via a serial port.
+ * <p/>
+ * This discovery service requires a component implementing the interface {@link UsbSerialDiscovery}, which performs the
+ * actual serial port and USB device discovery (as this discovery might differ depending on the operating system).
+ *
+ * @author Henning Sudbrock - initial contribution
+ */
+@NonNullByDefault
+@Component(immediate = true, service = { DiscoveryService.class,
+        UsbSerialDiscoveryService.class }, configurationPid = "discovery.usbserial")
+public class UsbSerialDiscoveryService extends AbstractDiscoveryService implements UsbSerialDiscoveryListener {
+
+    private final Logger logger = LoggerFactory.getLogger(UsbSerialDiscoveryService.class);
+
+    private final Set<UsbSerialDiscoveryParticipant> discoveryParticipants = new CopyOnWriteArraySet<>();
+
+    private final Set<UsbSerialDeviceInformation> previouslyDiscovered = new CopyOnWriteArraySet<>();
+
+    @NonNullByDefault({})
+    private UsbSerialDiscovery usbSerialDiscovery;
+
+    public UsbSerialDiscoveryService() {
+        super(5);
+    }
+
+    @Override
+    @Activate
+    protected void activate(@Nullable Map<String, @Nullable Object> configProperties) {
+        super.activate(configProperties);
+        usbSerialDiscovery.registerDiscoveryListener(this);
+        if (isBackgroundDiscoveryEnabled()) {
+            usbSerialDiscovery.startBackgroundScanning();
+        }
+    }
+
+    @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC)
+    protected void addUsbSerialDiscoveryParticipant(UsbSerialDiscoveryParticipant participant) {
+        this.discoveryParticipants.add(participant);
+        for (UsbSerialDeviceInformation usbSerialDeviceInformation : previouslyDiscovered) {
+            DiscoveryResult result = participant.createResult(usbSerialDeviceInformation);
+            if (result != null) {
+                thingDiscovered(result);
+            }
+        }
+    }
+
+    protected void removeUsbSerialDiscoveryParticipant(UsbSerialDiscoveryParticipant participant) {
+        this.discoveryParticipants.remove(participant);
+    }
+
+    @Reference
+    protected void setUsbSerialDiscovery(UsbSerialDiscovery usbSerialDiscovery) {
+        this.usbSerialDiscovery = usbSerialDiscovery;
+    }
+
+    protected synchronized void unsetUsbSerialDiscovery(UsbSerialDiscovery usbSerialDiscovery) {
+        usbSerialDiscovery.stopBackgroundScanning();
+        usbSerialDiscovery.unregisterDiscoveryListener(this);
+        this.usbSerialDiscovery = null;
+        this.previouslyDiscovered.clear();
+    }
+
+    @Modified
+    @Override
+    protected void modified(@Nullable Map<@NonNull String, @Nullable Object> configProperties) {
```

Comment:
could you please move that up, right below the `activate(...) method?

Reply:
Yes, will change.